sdk/src/main/java/com/google/cloud/dataflow/sdk/util/common/Counter.java

```diff
@@ -418,17 +430,44 @@ public boolean isCompatibleWith(Counter<?> that) {
 
   //////////////////////////////////////////////////////////////////////////////
 
-  /** The name of this counter. */
-  protected final String name;
-
   /** The kind of aggregation function to apply to this counter. */
   protected final AggregationKind kind;
 
   /** The commit state of this counter. **/
   protected final AtomicReference<CommitState> commitState;
 
+  /** A hashable name structure which will uniquely identify a counter. */
+  public interface Name {
+    public String name();
+    @Nullable public String systemName();
```

Comment:
I don't see us ever accessing the systemName/originalName of the counter. Could we do:

public interface Name {
  public String name();
}

public abstract static class StructuredName implemnets Name {
  abstract String contextSystemName();
  abstract String contextOriginalName();

  // static constructors
}
```

So that it isn't publicly exposed that there are different kinds of names?